### DISCUSSION TITLE
Heatmap of Grid-EYE on Pi Zero W

### DISCUSSION THREAD

**Discussion by Conclusio** [mKeRix/room-assistant#721] [Q&A] [answered]
I used the same Grid-EYE sensor on both Pi Zero W and Pi 3b, connected to Home Assistant. I don't experience obvious problems with the Pi 3b. It is a different story on the Pi Zero.

When enabling the heatmap on the Pi Zero it shows:
- "dead" pixels at random locations (sensor power at 3V or 5V - same effect)
  <img src="https://user-images.githubusercontent.com/10872142/119321361-da567d80-bc7c-11eb-97c1-f5fda4a4b667.png" width="256">
- close to 100% CPU usage
  <img src="https://user-images.githubusercontent.com/10872142/119322070-93b55300-bc7d-11eb-8168-070877e5ed9d.png" width="600">
- heatmap "reaction time" of more than 2 minutes

While the heatmap is refreshed every couple of seconds and shows changes in the temperature readings, it takes multiple minutes until the heatmap actually reflects the current situation. This can be tested by fully covering the sensor with your hand. From my point of view, it looks like there is a queue of sensor readings waiting to be converted to a heatmap.

### Instead of waddling through the code, here are my questions:
Is there a queue of sensor readings waiting to be processed?
And if so, can the size of this queue be configured?
A queue size of one would mean that only the latest sensor readings are used for the heatmap generation and older readings would be discarded.

**Comment by mKeRix** [comment, marked as answer]
There is no queue for generating heat maps atm, if new readings come in while the old ones haven't finished rendering yet a new render is simply launched in parallel (as per the async nature of JS). This is a simple implementation that doesn't really cause much issues if the image is rendered in time, but as you saw some boards like the Pi Zero W don't manage to do this with the current implementation.

The heatmap is the victim of a dependency change I did a while back. Initially I wanted to reduce the amount of native dependencies (i.e. things that need to be compiled) and also get rid of a memory leak caused by the previous dependency. I switched to a JavaScript-only library which worked similarly well, but it came with decreased performance. Over time it has shown that this decision wasn't great.

I currently have some changes locally that speed up image generation and should at least get rid of the delayed heatmap response for you. To achieve this I went back to the native dependencies (for their performance) and ensured that only a single render happens at a time. You can look out for this in the next release.

As for the "dead" pixels: I'm not sure where these are coming from - they might be caused by the high CPU usage. For the Grid-EYE each pixel is read separately from the I2C bus, so communication failures can express themselves through random pixels reporting weird values.

**Reply by Conclusio** [reply]
Thank you for the quick response!

> To achieve this I went back to the native dependencies (for their performance) and ensured that only a single render happens at a time. You can look out for this in the next release.

That sounds good. I'll keep track of this.
For now, I disabled the heat map and stick to the occupancy count. That's good enough for me atm.
